packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -2,104 +2,220 @@
-        external
-        virtual
-        AdjudicatorOnly
-    {
-        // no checks
-        //
-        // effects and interactions
-        _transfer(channelId, allocationBytes, new uint256[](0));
-    }
+        {
+            Outcome.OutcomeItem[] memory outcome = abi.decode(
+                guarantorOutcomeBytes,
+                (Outcome.OutcomeItem[])
+            );
+            Outcome.AssetOutcome memory assetOutcome = abi.decode(
+                outcome[assetIndex].assetOutcomeBytes,
+                (Outcome.AssetOutcome)
+            );
+            require(
+                assetOutcome.assetOutcomeType == Outcome.AssetOutcomeType.Guarantee,
+                '!guarantee'
+            );
+            guarantee = abi.decode(assetOutcome.allocationOrGuaranteeBytes, (Outcome.Guarantee));
+            asset = outcome[assetIndex].assetHolderAddress;
+        }
+        {
+            _requireChannelFinalized(guarantee.targetChannelId);
+            _requireMatchingFingerprint(
+                targetStateHash,
+                targetChallengerAddress,
+                keccak256(targetOutcomeBytes),
+                guarantee.targetChannelId
+            );
+        }
+        Outcome.AllocationItem[] memory allocation;
+        {
+            Outcome.OutcomeItem[] memory outcome = abi.decode(
+                targetOutcomeBytes,
+                (Outcome.OutcomeItem[])
+            );
+            Outcome.AssetOutcome memory assetOutcome = abi.decode(
+                outcome[assetIndex].assetOutcomeBytes,
+                (Outcome.AssetOutcome)
+            );
+            require(
+                assetOutcome.assetOutcomeType == Outcome.AssetOutcomeType.Allocation,
+                '!allocation'
+            );
+            allocation = abi.decode(
+                assetOutcome.allocationOrGuaranteeBytes,
+                (Outcome.AllocationItem[])
+            );
 
-    /**
-     * @notice Sets the given assetOutcomeHash for the given channelId in the assetOutcomeHashes storage mapping.
-     * @dev Sets the given assetOutcomeHash for the given channelId in the assetOutcomeHashes storage mapping.
-     * @param channelId Unique identifier for a state channel.
-     * @param assetOutcomeHash The keccak256 of the abi.encode of the Outcome.
-     */
-    function setAssetOutcomeHash(bytes32 channelId, bytes32 assetOutcomeHash)
-        external
-        AdjudicatorOnly
-    {
-        _setAssetOutcomeHash(channelId, assetOutcomeHash);
+            // effects and interactions
+            allocation = _claim(asset, guarantorChannelId, guarantee, allocation, indices);
+            outcome[assetIndex].assetOutcomeBytes = abi.encode(
+                Outcome.AssetOutcome(Outcome.AssetOutcomeType.Allocation, abi.encode(allocation))
+            );
+            bytes32 outcomeHash = keccak256(abi.encode(outcome));
+            _updateFingerprint(
+                guarantee.targetChannelId,
+                targetStateHash,
```

Comment:
Is it an issue that we update the outcome but do not update the state hash?

Reply:
I don't think so. As long as it is clear that the `stateHash` is a record of the latest state the chain is aware of, and it can only change (on chain) *before* the channel finalizes. After finalization, the `outcome` can change (on chain) without the `stateHash` changing. 

Of course the `stateHash` refers to an outcome, and one might think that implies some redundancy -- why not simply update the outcome in the `stateHash`? I think what we have now is more efficient, because we only have to supply the `stateHash` and the preimage of the `outcomeHash` in the calldata for `transfer`. If we only stored that `stateHash` and applied transformations to that directly, we would need to supply the entire preimage of the `stateHash` every time.